src/aiokatcp/sensor.py

```diff
@@ -714,3 +715,69 @@ def copy(self) -> Dict[str, Sensor]:
     values.__doc__ = dict.values.__doc__
     items.__doc__ = dict.items.__doc__
     copy.__doc__ = dict.copy.__doc__
+
+
+# I'm ever so slightly nervous about inheritance and having a metaclass.
+class AggregateSensor(Sensor, metaclass=ABCMeta):
+    """A Sensor with its reading determined by several other Sensors.
+
+    This is an abstract class, the user must implement :meth:`_update_aggregate`
+    in order to define the logic of how the aggregate sensor's reading will be
+    determined by those of the sensor set to which it is targeted.
+
+    Parameters are all as per :class:`Sensor`, with the exception of
+    `target`, which is the :class:`SensorSet` from which the aggregated sensor
+    will determine its own reading.
+    """
+
+    def __init__(
+        self,
+        target: SensorSet,
+        sensor_type: Type[_T],
+        name: str,
+        description: str = "",
+        units: str = "",
+        default: _T = None,
+        initial_status: Sensor.Status = Sensor.Status.UNKNOWN,
+        *,
+        status_func: Callable[[_T], Sensor.Status] = _default_status_func,
+        auto_strategy: Optional["SensorSampler.Strategy"] = None,
+        auto_strategy_parameters: Iterable[Any] = (),
+    ) -> None:
+        Sensor.__init__(
+            self,
+            sensor_type=sensor_type,
+            name=name,
+            description=description,
+            units=units,
+            default=default,
+            initial_status=initial_status,
+            status_func=status_func,
+            auto_strategy=auto_strategy,
+            auto_strategy_parameters=auto_strategy_parameters,
+        )
+        self.target = target
+        for sensor in self.target.values():
+            if sensor is not self:  # This sensor might be in the set.
+                sensor.attach(self._update_aggregate)
+
+        self.target.add_add_callback(self._sensor_added)
+        self.target.add_remove_callback(self._sensor_removed)
+
+    @abstractmethod
+    def _update_aggregate(self, updated_sensor: Sensor, reading: Reading):
+        return
+
+    def _sensor_added(self, sensor: Sensor):
+        """Add the update callback to a new sensor in the set."""
+        if sensor is not self:
+            sensor.attach(self._update_aggregate)
+        # The new sensor might actually affect the aggregate reading.
+        self._update_aggregate(sensor, sensor.reading)
+
+    def _sensor_removed(self, sensor: Sensor):
+        """Remove the update callback from a sensor no longer in the set."""
+        if sensor is not self:
+            sensor.detach(self._update_aggregate)
+        # The new sensor might actually affect the aggregate reading.
```

Comment:
You have a point. In my use case, it doesn't really matter, because the callback only looks at the sensors in `self.target`, and the actual argument of the callback is ignored. But that won't be true in the most general case.

Reply:
Or will it? Is there a case where a newly-added or removed sensor may have special status over the others in the set for determining the aggregate?

In my mind the fact that the set of sensors has changed (in any way, either addition or removal) should just trigger a re-calculation of the aggregate value / status from the sensors in the set post-change.

Reply:
There may be cases where an aggregate can be updated significantly more cheaply by using information about what's changed. For example, if you have 10000 sensors and your aggregate counts how many of them are in error, then checking one sensor change to increment or decrement the count would be cheaper than checking all 10000 sensors any time any of the 10000 sensors changes.

To make that work, how about passing:
- the sensor object
- the old reading, or None if it's an addition
- the new reading, or None if it's a removal

I haven't re-read the code for a while so I'm not sure how accessible the old value is when the sensor value is updated.

Reply:
I suggest passing `None, None, None` when the aggregate is constructed, since that's a fourth case.